README.rst
Overview
===========================

kanyun是一个openstack的监控和度量工具， 他可以跟踪租户的资源使用情况，包括虚拟机的CPU使用率、
内存使用、网络流量和磁盘读写情况，同时可以进行数据汇总和统计。
kanyun包括worker、data-server、API-server三个模块。
worker模块负责采集监控数据并交由data-server进行存储，APIserver则提供进行查询和统计的外部接口。

kanyun采集的数据可以由dashboard进行展示，或者由计费系统（如dough）计费使用。

拓扑
----

总览::

                    +-----------------+
                    |    cassandra    |
                    +-----------------+
                      |            |
                      |            |
                +-----------+ +-------------+
                |   server  | |  kanyun-API |   <---- client 
                +-----------+ +-------------+
                       | 
                       |    
                     woker
                       |   
               ________|___________        
             /         |           \
     外网流量采集   虚拟机信息采集  其他采集模块 

Server::


                         +--- <-- Worker's PUSH
                         |
                         |
                    +----------+
                    |   PULL   |     <-- feedback
                +---|==========|
    Client--->  |REP|  Server  |
                +---|==========|
                    |   PUB    |     <-- broadcast
                    +----------+
                         |
                         |
                         +----> Worker's SUB
                         +----> DB

部署::

安装cassandra数据库
安装ZeroMQ
创建和配置/etc/kanyun.conf
在服务器上启动kanyun-api和kanyun-server
在nova-compute节点上启动kanyun-worker

通讯协议
------

数据通讯全部使用json格式


外网流量采集模块--->worker
------------------------

格式::

    {'instanceid1':('IP', time, 'value'), 'instanceid2':('IP', time, 'value')}
    value的单位：KB


示例::

    {'instance-00000001': ('10.0.0.2', 1332409327, '0')}

虚拟机信息采集模块-->worker


格式::

    [('cpu', 'total', (utc_time, cpu_usage)), 
    ('mem', 'total', (utc_time, max, free)), 
    ('nic', 'vnet8', (utc_time, incoming, outgoing(内网))), 
    ('blk', 'vda', (utc_time, read, write)), 
    ('blk', 'vdb', (utc_time, read, write))],

示例::

    {'instance-000000ba@sws-yz-5': 
    [('cpu', 'total', (1332400088.149444, 5)),  
    ('mem', 'total', (1332400088.149444, 8388608L, 8388608L)), 
    ('nic', 'vnet8', (1332400088.152285, 427360746L, 174445810L)), 
    ('blk', 'vda', (1332400088.156346, 298522624L, 5908590592L)), 
    ('blk', 'vdb', (1332400088.158202, 2159616L, 1481297920L))], 
    其中cpu和mem为实际使用量

worker-->server
----------------

格式::

    ['msg_type', data]

    'msg_type'取值：
        HEART_BEAT = '0'
        LOCAL_INFO = '1'
        TRAFFIC_ACCOUNTING = '2'
        AGENT = '3'

示例

心跳::
    ['WORKER1', time.time(), status]
    status：0为即将正常退出，服务器收到0就会取消对该worker的状态监控；1为正在工作中
    数据：
    ['2', 
        ''{"instance-00000001@pyw.novalocal": 
            [
             ["cpu", "total", [1332831360.029795, 53522870000000]], 
             ["mem", "total", [1332831360.029795, 131072, 131072]], 
             ["nic", "vnet0", [1332831360.037399, 21795245, 5775663]], 
             ["blk", "vda", [1332831360.04699, 474624, 4851712]], 
             ["blk", "vdb", [1332831360.049333, 122880, 0]]
            ]
         }''
    ]

server-->cassandra
--------------------

格式::

    instance_id, {scf_str: {时间: 值}}

示例::

    Columnfamily为mem_max保存最大内存值，mem_free保存空闲内存值
    instance_id, {'total': {1332831360: 131072}}
    instance_id, {'total': {1332831360: 131072}}

billing -->API server
--------------------

格式::

    ['msg_type', 'uuid', {
        'method': 'query_usage_report',
        'args':  {
                'id': 'instance00001'
                'metric': 'network',
                'metric_param': 'vnet0',
                'statistic': 'sum',
                'period': 5,
                'timestamp_from': '2012-02-20T12:12:12',
                'timestamp_to': '2012-02-22T12:12:12',
            }
        }
    ]
    metric取值：

    'vmnetwork'
    'cpu'
    'mem_max'
    'mem_free'
    'nic_read'
    'nic_write'
    'blk_read'
    'blk_write'
    metric_param取值：

    如果metric为vmnetwork或者cpu或者mem*的话，为'total'
    其他情况为设备名，比如'vnet0'、'vda'等
    statistic取值:

    'sum'
    'max'
    'min'
    'avg'
    'sam'(暂时不支持)

示例

协议请求数据示例::

    ['msg_type', 'uuid', {
        'method': 'query_usage_report',
        'args':  {
                'id': 'instance00001'
                'metric': 'vmnetwork',
                'metric_param': 'vnet0',
                'statistic': 'sum',
                'period': 5,
                'timestamp_from': '2012-02-20T12:12:12',
                'timestamp_to': '2012-02-22T12:12:12',
            }
        }
    ]
api_client示例::

    api-client instance-00000001@pyw.novalocal cpu total sum 5 2012-02-20T12:12:12 2012-06-20T12:12:12

    获取存在指定数据的全部实例列表： 
    api_client vmnetwork 
    获取指定实例的数据： 
    api_client -k instance-0000002 
    获取指定类型、指定实例、指定参数的数据： 
    api_client instance-0000002 vmnetwork 10.0.0.2 api_client instance-00000012@lx12 cpu api_client instance-00000012@lx12 mem mem_free 
    查询指定实例、指定类型、指定参数、指定统计类型的数据，以5分钟为统计单位、从指定时间开始到当前时间进行统计，返回统计结果： 
    api_client instance-0000002 vmnetwork 10.0.0.2 0 5 1332897600 0


API server--> billing
--------------------

格式::

    ['msg_type', 'uuid',
        {'code': 0,
         'message': 'success',
         'data':{key:result},
        }
    ]

示例::

    [ {"1332897600.0": 10} ]

数据库


结构::

    +--------------+
    | cf=vmnetwork |
    +--------------+-------------------------------------------+
    | scf=IP                                                   |
    +===================+===========+=======+==================+
    |                   | col=time1 | time2 | ...              |
    +===================+===========+=======+==================+
    | key=instance_id   |   val1    | val2  | ...              |
    +==========================================================+

    +---------------------------------------------------------------------------------------------+
    | cf=cpu/mem_max/mem_free/nic_read/nic_write/blk_read/blk_write/...(one item as one cf )      |
    +---------------------------------------------------------------------------------------------+
    | scf=total/devname(vnet0/vda...)                  |
    +=================+==============+=======+=========+
    |                 | col=utc_time | time2 | ...     |
    +=================+==============+=======+=========+
    | key=instance_id | val1(subval) | val2  | ...     |
    +==================================================+

建库
----

可以在数据库本地使用cassandra-cli -h 127.0.0.1连接数据库并执行以下命令建库::

    CREATE keyspace DATA;
    USE DATA;
     
    CREATE COLUMN family vmnetwork WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family cpu WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family mem_max WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family mem_free WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family nic_incoming WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family nic_outgoing WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family blk_read WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
    CREATE COLUMN family blk_write WITH column_type='Super' AND comparator='AsciiType' AND subcomparator='IntegerType' AND default_validation_class='AsciiType';
     
    assume vmnetwork KEYS AS ascii;
    assume cpu KEYS AS ascii;
    assume mem_max KEYS AS ascii;
    assume nic_incoming KEYS AS ascii;
    assume nic_outgoing KEYS AS ascii;
    assume blk_read KEYS AS ascii;
    assume blk_write KEYS AS ascii;
    assume mem_free KEYS AS ascii;

schema::

    [DEFAULT@DATA] SHOW schema;
    CREATE keyspace DATA
      WITH placement_strategy = 'NetworkTopologyStrategy'
      AND strategy_options = {datacenter1 : 1}
      AND durable_writes = true;
     
    USE DATA;
     
    CREATE COLUMN family blk_read
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family blk_write
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family cpu
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family mem_free
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family mem_max
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family nic_incoming
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family nic_outgoing
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family nic_read
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family nic_write
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';
     
    CREATE COLUMN family vmnetwork
      WITH column_type = 'Super'
      AND comparator = 'AsciiType'
      AND subcomparator = 'IntegerType'
      AND default_validation_class = 'AsciiType'
      AND key_validation_class = 'BytesType'
      AND rows_cached = 0.0
      AND row_cache_save_period = 0
      AND row_cache_keys_to_save = 2147483647
      AND keys_cached = 200000.0
      AND key_cache_save_period = 14400
      AND read_repair_chance = 1.0
      AND gc_grace = 864000
      AND min_compaction_threshold = 4
      AND max_compaction_threshold = 32
      AND replicate_on_write = true
      AND row_cache_provider = 'SerializingCacheProvider'
      AND compaction_strategy = 'org.apache.cassandra.db.compaction.SizeTieredCompactionStrategy';

配置文件样例::

    bin/kanyun.conf
    [log]
    file=/tmp/kanyun.log
     
    [server]
    host: *
    port: 5551
    db_host: 127.0.0.1
     
    [api]
    api_host: *
    api_port: 5556
    db_host: 127.0.0.1
     
    [worker]
    id: worker1
    worker_timeout: 60
    dataserver_host: 127.0.0.1
    dataserver_port: 5551
    log: /tmp/kanyun-worker.log
     
    [client]
    api_host: 127.0.0.1
    api_port: 5556


debian/changelog
kanyun (0.1.1-1) unstable; urgency=low

  * Initial release (Closes: #nnnn)  <nnnn is the bug number of your ITP>

 -- Tang Yaguang <yaguang@staff.sina.com.cn>  Tue, 24 Jul 2012 10:53:31 +0800


kanyun/server/api_server.py
# vim: tabstop=4 shiftwidth=4 softtabstop=4
#
# Copyright 2012 Sina Corporation
# All Rights Reserved.
# Author: YuWei Peng <pengyuwei@gmail.com>
#
#    Licensed under the Apache License, Version 2.0 (the "License"); you may
#    not use this file except in compliance with the License. You may obtain
#    a copy of the License at
#
#         http://www.apache.org/licenses/LICENSE-2.0
#
#    Unless required by applicable law or agreed to in writing, software
#    distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
#    WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
#    License for the specific language governing permissions and limitations
#    under the License.

import sys
import time
import types
import json
import traceback
import ConfigParser
import zmq
from dateutil.parser import parse
from collections import OrderedDict

from nova import utils
from kanyun.common.const import *
from kanyun.common.buffer import HallBuffer
from kanyun.database.cassadb import CassaDb

"""
Save the vm's system info data to db.

protocol:
    http://wiki.sinaapp.com/doku.php?id=monitoring
[u'S', u'instance-000001@pyw.novalocal', u'cpu', u'total', 0, 5, 1332897600, 0]
"""

class Statistics():

    def __init__(self):
        self.clean()
        
    def clean(self):
        self.first = True
        self.count = 0
        self.sum = 0.0
        self.min = 0.0
        self.max = 0.0
        self.previous = 0.0
        self.diff = 0.0
        
    def update(self, value):
        self.count += 1
        self.sum += value
        if self.first:
            self.first = False
            self.previous = value
            self.max = value
            self.min = value
            return
            
        if value > self.previous:
            self.max = value
        elif value < self.previous:
            self.min = value
        self.diff = value - self.previous
        self.previous = value
        
    def get_value(self, w):
        if w == 'avg' or w == STATISTIC.AVERAGE:
            return self.get_agerage()
        elif w == 'min' or w == STATISTIC.MINIMUM:
            return self.get_min()
        elif w == 'max' or w == STATISTIC.MAXIMUM:
            return self.get_max()
        elif w == 'sum' or w == STATISTIC.SUM:
            return self.get_sum()
        elif w == 'sam' or w == STATISTIC.SAMPLES:
            return self.get_samples()
        else:
            print 'error:', w
            return 0
            
    def get_diff(self):
        return self.diff
        
    def get_agerage(self):
        if self.count == 0:
            return 0
        else:
            return self.sum / self.count
    def get_sum(self):
        return self.sum
        
    def get_max(self):
        return self.max
        
    def get_min(self):
        return self.min
        
    def get_samples(self):
        # TODO
        return 0;


class ApiServer():

    def __init__(self, 
                 db_host='127.0.0.1', 
                 log_file="/tmp/api-server.log",
                 log_level=logging.NOTSET):
        # cassandra database object
        self.db = None
        self.db_host = db_host
        self.buf = HallBuffer()
        
    def get_db(self):
        if self.db is None:
            self.db = CassaDb('data', self.db_host)
        return self.db
        
    def get_data(self, row_id, cf_str, scf_str, time_from=0, time_to=0):
        """
        param type: UnicodeType and IntType
        return: recordset, count, bool(count > limit?)
        """
        if not isinstance(row_id, unicode) \
            or not isinstance(cf_str, unicode) \
            or not isinstance(scf_str, unicode) \
            or not isinstance(time_from, int) \
            or not isinstance(time_to, int):
            return None, 0, True
            
        if time_to == 0:
            time_to = int(utils.utcnow_ts())
        print "get_data:", row_id, cf_str, scf_str, time_from, time_to
        
        bufkey = str([row_id, cf_str, scf_str, time_from, time_to])
        if self.buf.hit_test(bufkey):
            return self.buf.get_buf(bufkey)
        
        db = self.get_db()
        rs = db.get(cf_str, row_id, super_column=scf_str, 
                column_start=time_from, column_finish=time_to, column_count=20000)
        count = 0 if rs is None else len(rs)
        
        ret = rs, count, False if (count == 20000) else True
        if not rs is None:
            self.buf.save(bufkey, ret)
        
        return ret
        
    def analyize_data(self, rs, period, statistic, show=False):
        """[private func]analyize the data
        period: minutes
        return: {"key":"value"}
        note: timestmp is already UTC time, so do NOT use time.gmtime again!
        """
        if rs is None \
            or not isinstance(period, int):
            return None
        t = 0
        key_time = 0
        st = Statistics()
        this_period = dict()
        
        for timestmp, value in rs.iteritems():
            rt = time.localtime(timestmp)
            key = rt.tm_min + rt.tm_hour*100 + rt.tm_mday*10000 + \
                  rt.tm_mon*1000000 + rt.tm_year*100000000
            if t == 0:
                st.clean()
                t = timestmp
                key_time = time.localtime(timestmp)
            if timestmp >= t + period*60:
                st.clean()
                t = timestmp
                key_time = time.localtime(timestmp)
            st.update(float(value))
            key2 = time.mktime(
                              (key_time.tm_year, key_time.tm_mon, key_time.tm_mday,
                              key_time.tm_hour, key_time.tm_min,0,0,0,0))
            this_period[key2] = st.get_value(statistic)
                
        this_period = OrderedDict(sorted(this_period.items(), key=lambda t: t[0]))
        if show:
            print statistic, ":each period(", period, "):"
        for m, val in this_period.iteritems():
            rt = time.localtime(m)
            key = rt.tm_min + rt.tm_hour*100 + rt.tm_mday*10000 + \
                  rt.tm_mon*1000000 + rt.tm_year*100000000
            if show:
                print '\t', key, m, val
            
        return this_period

    ######################### public API interface ########################
    def get_instances_list(self, cf_str):
        if not isinstance(cf_str, unicode):
            print 'param types error'
            return None
        ret = list()
        limit = 20000
        time_to = int(utils.utcnow_ts())
        time_from = time_to - 24 * 60 * 60
        db = self.get_db()
        
        rs = db.get_range2(cf_str, row_count=20)
        return list(rs)
        if not rs is None:
            for i in rs:
                ret.append(i[0])
        
        return ret
        
    def get_by_instance_id(self, row_id, cf_str):
        if not isinstance(row_id, unicode) \
            or not isinstance(cf_str, unicode):
            print 'param types error'
            return None, 0, True
        db = self.get_db()
        rs = db.getbykey(cf_str, row_id)
        count = 0 if rs is None else len(rs)
        
        return rs, count, False if (count == 20000) else True
        
    def get_by_key(self, row_id, cf_str, scf_str, limit=20000):
        """
        example:cf=u'vmnetwork',scf=u'10.0.0.1',key=u'instance-0000002'
        return: recordset, count, bool(count > limit?)
        """
        if not isinstance(row_id, unicode) \
            or not isinstance(cf_str, unicode) \
            or not isinstance(scf_str, unicode) \
            or not isinstance(limit, int):
            print 'param types error'
            return None, 0, True
        db = self.get_db()
        rs = db.getbykey2(cf_str, key=row_id, super_column=scf_str, column_count=limit)
        count = 0 if rs is None else len(rs)
        
        return rs, count, False if (count == 20000) else True

    def query_usage_report(self, args, **kwargs):
# TODO: how to use kwargs?
#    def query_usage_report(self, arg, id=None, metric='cpu', 
#                           metric_param='total',
#                           statistic='avg', period=5,
#                           timestamp_from=None, timestamp_to=None,
#                           **kwargs):
        """statistic is STATISTIC enum
        period default=5 minutes
        time_to default=0(now)"""
        """
        {
            'id': 'instance00001',
            'metric': 'network',
            'metric_param': 'vnet0',
            'statistic': 'sum',
            'period': 5,
            'timestamp_from': '2012-02-20T12:12:12',
            'timestamp_to': '2012-02-22T12:12:12',
        }
        """
#        usage_report = dict()
#        datetime_from = iso8601.parse_date(timestamp_from)
#        datetime_to = iso8601.parse_date(timestamp_to)
#        # TODO: implement
#        return {'data': usage_report}

        row_id = args['id']
        cf_str = args['metric']
        scf_str = args.setdefault("metric_param", "total")
        statistic = args['statistic']
        period = int(args['period'])
        timestamp_from = args['timestamp_from']
        timestamp_to = args['timestamp_to']
#        time_from = iso8601.parse_date(timestamp_from)
#        time_from = int(time.mktime(time_from.timetuple()))
        time_from = parse(timestamp_from)
        time_from = int(time.mktime(time_from.timetuple()))
        time_to = int(utils.utcnow_ts())
        if not timestamp_to is None:
#            time_to = iso8601.parse_date(timestamp_to)
#            time_to = int(time.mktime(time_to.timetuple()))
            time_to = parse(timestamp_to)
            time_to = int(time.mktime(time_to.timetuple()))
            
        bufkey = str([row_id, cf_str, scf_str, 
                      statistic, period, time_from, time_to])
        print "query_usage_report time range:", time_from, time_to
        if self.buf.hit_test(bufkey):
            print "buffer hit:", bufkey
            return self.buf.get_buf(bufkey)
            
        ret_len = 0
        (rs, count, all_data) = self.get_data(row_id, cf_str, scf_str, 
                                          time_from, time_to)
        if not rs is None and count > 0:
            buf = self.analyize_data(rs, 1, statistic)
            ret = self.analyize_data(buf, period, statistic, show=True)
            if ret is None:
                ret_len = 0
            else:
                ret = OrderedDict(sorted(ret.items(), key=lambda t: t[0]))
                ret_len = len(ret)
            print ret_len, "result."
        else:
            print "no result."
            ret = None
            ret_len = 0
            
        result = ret, ret_len, all_data
        if (not result is None and utils.utcnow_ts() - time_to > 120):
            self.buf.cleanup()
            self.buf.save(bufkey, result)
        return result


    def statistic(self, row_id, cf_str, scf_str, 
                      statistic, period=5, time_from=0, time_to=0):
        """statistic is STATISTIC enum
        period default=5 minutes
        time_to default=0(now)"""
        if (not isinstance(row_id, unicode) \
            or not isinstance(cf_str, unicode) \
            or not isinstance(scf_str, unicode) \
            or not isinstance(statistic, basestring) \
            or not isinstance(period, int) \
            or not isinstance(time_from, int) \
            or not isinstance(time_to, int)):
            print 'param types error'
            return None, 0, True
            
        bufkey = str([row_id, cf_str, scf_str, 
                      statistic, period, time_from, time_to])
        if self.buf.hit_test(bufkey):
            return self.buf.get_buf(bufkey)
            
        ret_len = 0
        (rs, count, all_data) = self.get_data(row_id, cf_str, scf_str, 
                                          time_from, time_to)
        if not rs is None and count > 0:
            buf = self.analyize_data(rs, 1, statistic)
            ret = self.analyize_data(buf, period, statistic, show=True)
            if ret is None:
                ret_len = 0
            else:
                ret = OrderedDict(sorted(ret.items(), key=lambda t: t[0]))
                ret_len = len(ret)
            print ret_len, "result."
        else:
            print "no result."
            ret = None
            ret_len = 0
            
        result = ret, ret_len, all_data
        self.buf.save(bufkey, result)
        return result

    ##################### end public API interface ########################

